crates/ty/src/printer.rs

```diff
@@ -0,0 +1,130 @@
+use std::io::StdoutLock;
+
+use indicatif::ProgressDrawTarget;
+
+use crate::logging::VerbosityLevel;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub(crate) struct Printer {
+    verbosity: VerbosityLevel,
+    no_progress: bool,
+}
+
+impl Printer {
+    #[must_use]
+    pub(crate) fn with_no_progress(self) -> Self {
+        Self {
+            verbosity: self.verbosity,
+            no_progress: true,
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn with_verbosity(self, verbosity: VerbosityLevel) -> Self {
+        Self {
+            verbosity,
+            no_progress: self.no_progress,
+        }
+    }
+
+    /// Return the [`ProgressDrawTarget`] for this printer.
+    pub(crate) fn progress_target(self) -> ProgressDrawTarget {
+        if self.no_progress {
+            return ProgressDrawTarget::hidden();
+        }
+
+        match self.verbosity {
+            VerbosityLevel::Quiet => ProgressDrawTarget::hidden(),
+            VerbosityLevel::Default => ProgressDrawTarget::stderr(),
+            // Hide the progress bar when in verbose mode.
+            // Otherwise, it gets interleaved with log messages.
+            VerbosityLevel::Verbose => ProgressDrawTarget::hidden(),
+            VerbosityLevel::ExtraVerbose => ProgressDrawTarget::hidden(),
+            VerbosityLevel::Trace => ProgressDrawTarget::hidden(),
+        }
+    }
+
+    /// Return the [`Stdout`] stream for an important message.
+    pub(crate) fn stdout_important(self) -> Stdout {
+        match self.verbosity {
+            VerbosityLevel::Quiet => Stdout::enabled(),
+            VerbosityLevel::Default => Stdout::enabled(),
+            VerbosityLevel::Verbose => Stdout::enabled(),
+            VerbosityLevel::ExtraVerbose => Stdout::enabled(),
+            VerbosityLevel::Trace => Stdout::enabled(),
+        }
+    }
+
+    /// Return the [`Stdout`] stream for general messages.
+    pub(crate) fn stdout(self) -> Stdout {
+        match self.verbosity {
+            VerbosityLevel::Quiet => Stdout::disabled(),
+            VerbosityLevel::Default => Stdout::enabled(),
+            VerbosityLevel::Verbose => Stdout::enabled(),
+            VerbosityLevel::ExtraVerbose => Stdout::enabled(),
+            VerbosityLevel::Trace => Stdout::enabled(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum StreamStatus {
+    Enabled,
+    Disabled,
+}
+
+#[derive(Debug)]
+pub(crate) struct Stdout {
+    status: StreamStatus,
+    lock: Option<StdoutLock<'static>>,
+}
+
+impl Stdout {
+    pub(crate) fn lock(mut self) -> Self {
+        if self.status == StreamStatus::Enabled {
+            self.lock = Some(std::io::stdout().lock());
+        } else {
+            self.lock = None;
+        }
+        self
+    }
```

Comment:
We don't support locking in uv's printer, but in ty we lock stdout for the hot loop of printing diagnostics which seems reasonable and worth retaining.